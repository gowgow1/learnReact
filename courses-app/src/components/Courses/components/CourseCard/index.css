.card {
	display: flex;
	column-gap: 20px;
	border: 2px solid rgb(120, 240, 22);
	padding: 15px;
	margin-bottom: 20px;
}

.card-content {
	width: 70%;
}
.card-info {
	width: 30%;
}

.title {
	font-size: 20px;
	margin:0;
	margin-bottom: 20px;
}

.author,
.duration,
.created {
	margin-bottom: 15px;
}
.btn-wrap {
	display: flex;
	align-items: center;
	justify-content: center;
	flex-wrap: wrap;
	row-gap: 5px;
	column-gap: 5px;
}
.btn-wrap .button {
	border: 3px solid plum;
	padding: 5px 25px;
}