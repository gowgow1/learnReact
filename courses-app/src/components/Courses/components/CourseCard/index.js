import CourseCards from './CourseCard';
export default CourseCards;
