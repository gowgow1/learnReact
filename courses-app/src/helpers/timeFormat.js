const timeFormat = (string) => {
	const hours = '' + Math.trunc(+string / 60);
	const min = '' + (+string % 60);
	return string ? `${hours.padStart(2, '0')}:${min.padStart(2, '0')}` : '00:00';
};

export default timeFormat;
