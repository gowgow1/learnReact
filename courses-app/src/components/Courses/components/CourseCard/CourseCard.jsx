import Button from '../../../common/Button';
import timeFormat from '../../../../helpers/timeFormat';

import './index.css';
import { useHistory } from 'react-router';
import { useDispatch } from 'react-redux';
import { deleteCourse } from '../../../../store/courses/actionCreators';

const CourseCards = ({
	id,
	title,
	description,
	author = [],
	duration,
	created,
}) => {
	const history = useHistory();
	const onClick = () => history.push(`/courses/${id}`);
	const dispatch = useDispatch();

	const onDelete = () => {
		dispatch(deleteCourse(id));
	};
	const onEdit = () => {}; //tasks of next hw

	return (
		<div className='card'>
			<div className='card-content'>
				<h2 className='title'>{title}</h2>
				<div className='description'>{description}</div>
			</div>
			<div className='card-info'>
				<div className='author'>
					<strong>Authors: </strong>
					{author.join(', ')}
				</div>
				<div className='duration'>
					<strong>Duration: </strong>
					{timeFormat(duration)}
				</div>
				<div className='created'>
					<strong>Created: </strong>
					{created}
				</div>
				<div className='btn-wrap'>
					<Button text='Show course' onClick={onClick} />
					<Button text='E' onClick={onEdit} />
					<Button text='D' onClick={onDelete} />
				</div>
			</div>
		</div>
	);
};

export default CourseCards;
