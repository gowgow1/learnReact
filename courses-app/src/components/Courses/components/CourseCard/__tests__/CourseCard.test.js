import * as reactRedux from 'react-redux';
import { render } from '@testing-library/react';
import CourseCard from '../CourseCard';
import { Provider } from 'react-redux';
import timeFormat from '../../../../../helpers/timeFormat';

const mockedState = {
	isAuth: true,
	name: 'Alex',
	token: '123123',
	role: 'user',
	email: 'alex@test.ru',
};

const mockedCourse = {
	id: 123,
	title: 'Course',
	description: 'description',
	created: '12/12/2021',
	duration: 123,
	author: ['Pau', 'Pau'],
};

const mockedStore = {
	getState: () => mockedState,
	subscribe: jest.fn(),
	dispatch: jest.fn(),
};

jest.mock('react-redux', () => ({
	...jest.requireActual('react-redux'),
	useSelector: jest.fn(),
}));

jest.spyOn(reactRedux, 'useDispatch');

beforeEach(() => {
	reactRedux.useSelector.mockReturnValue(mockedState);
});
afterEach(() => {
	reactRedux.useSelector.mockClear();
});

describe('Course Card', () => {
	it('should render', () => {
		const { container } = render(
			<Provider store={mockedStore}>
				<CourseCard {...mockedCourse} />
			</Provider>
		);
		expect(container).toMatchSnapshot();
	});

	it('CourseCard has title', () => {
		const { getByText } = render(
			<Provider store={mockedStore}>
				<CourseCard {...mockedCourse} />
			</Provider>
		);
		expect(getByText('Course')).toBeTruthy();
	});

	it('CourseCard has description', () => {
		const { getByText } = render(
			<Provider store={mockedStore}>
				<CourseCard {...mockedCourse} />
			</Provider>
		);
		expect(getByText('description')).toBeTruthy();
	});

	it('CourseCard has duration in the correct format', () => {
		const { getByText } = render(
			<Provider store={mockedStore}>
				<CourseCard {...mockedCourse} />
			</Provider>
		);
		const d = 'Duration: ' + timeFormat(mockedCourse.duration);
		expect(getByText(d)).toBeTruthy();
	});

	it('CourseCard has authors list', () => {
		const { getByText } = render(
			<Provider store={mockedStore}>
				<CourseCard {...mockedCourse} />
			</Provider>
		);
		expect(getByText(mockedCourse.author.join(', '))).toBeTruthy();
	});

	it('CourseCard has created Date', () => {
		const { getByText } = render(
			<Provider store={mockedStore}>
				<CourseCard {...mockedCourse} />
			</Provider>
		);
		expect(getByText('Created: ' + mockedCourse.created)).toBeTruthy();
	});
});
