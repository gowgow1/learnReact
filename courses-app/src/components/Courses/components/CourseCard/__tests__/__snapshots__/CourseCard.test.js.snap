// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Course Card should render 1`] = `
<div>
  <div
    class="card"
    data-testid="card"
  >
    <div
      class="card-content"
    >
      <h2
        class="title"
      >
        Course
      </h2>
      <div
        class="description"
      >
        description
      </div>
    </div>
    <div
      class="card-info"
    >
      <div
        class="author"
      >
        <strong>
          Authors: 
        </strong>
        Pau, Pau
      </div>
      <div
        class="duration"
      >
        Duration: 
        02:03
      </div>
      <div
        class="created"
      >
        Created: 
        12/12/2021
      </div>
      <div
        class="btn-wrap"
      >
        <button
          class="button"
        >
          Show course
        </button>
      </div>
    </div>
  </div>
</div>
`;
